src/envoy/http/path_matcher/filter_config.h

```diff
@@ -65,35 +66,45 @@ class FilterConfig : public Envoy::Logger::Loggable<Envoy::Logger::Id::filter> {
   }
 
   // Returns whether an operation needs path parameter extraction.
+  // If needed, it will also return a map of snake to json segment conversions.
+  //
   // NOTE: path parameter extraction is only needed when backend rule path
   // translation is CONSTANT_ADDRESS.
-  bool needParameterExtraction(const std::string& operation) const {
-    auto operation_it = path_params_operations_.find(operation);
-    return operation_it != path_params_operations_.end();
+  const ::espv2::api::envoy::v6::http::path_matcher::
+      PathParameterExtractionRule*
+      needParameterExtraction(const std::string& operation) const {
+    auto operation_it = path_param_extractions_.find(operation);
+    if (operation_it == path_param_extractions_.end()) {
+      return nullptr;
+    }
+
+    // Relies on pointer safety in absl::flat_hash_map.
+    // The map will never change after filter config is generated, so it should
+    // be safe.
+    return &(operation_it->second);
   }
 
   FilterStats& stats() { return stats_; }
 
-  // Returns the mapp from snake-case segment name to JSON name.
-  const absl::flat_hash_map<std::string, std::string>& getSnakeToJsonMap() {
-    return snake_to_json_map_;
-  }
-
  private:
   FilterStats generateStats(const std::string& prefix,
                             Envoy::Stats::Scope& scope) {
     const std::string final_prefix = prefix + "path_matcher.";
-    return {ALL_BACKEND_AUTH_FILTER_STATS(
+    return {ALL_PATH_MATCHER_FILTER_STATS(
         POOL_COUNTER_PREFIX(scope, final_prefix))};
   }
 
   ::espv2::api::envoy::v6::http::path_matcher::FilterConfig proto_config_;
   ::espv2::api_proxy::path_matcher::PathMatcherPtr<const std::string*>
       path_matcher_;
-  // Mapping between snake-case segment name to JSON name as specified in
-  // `Service.types` (e.g. "foo_bar" -> "fooBar").
-  absl::flat_hash_map<std::string, std::string> snake_to_json_map_;
-  absl::flat_hash_set<std::string> path_params_operations_;
+
+  // Map from operation id to a PathParameterExtractionRule.
+  // Only stores the operations that need path param extraction.
+  absl::flat_hash_map<
+      std::string,
+      ::espv2::api::envoy::v6::http::path_matcher::PathParameterExtractionRule>
```

Comment:
If we store the whole config_pb in this object, we don't copy each Rule message into the map, we can just store the reference to it

Reply:
Good point. Should not store references in a map. I'll change to pointers.